File: AI_Agent/functions/src/schemas.ts
import { z } from "zod";

// Parsed Input Schemas

export const CoordinatesSchema = z.object({
    lat: z.number().describe("Latitude of the location"),
    lon: z.number().describe("Longitude of the location"),
    place_name: z.string().describe("Human-readable name of this place"),
});

export const ObjectiveSchema = z
    .enum(["time", "cost", "eco"])
    .describe(
        "Optimization goal: 'time' = fastest, 'cost' = cheapest, 'eco' = lowest carbon emissions"
    );

export const ParsedSingleRouteSchema = z.object({
    intent: z.literal("single_route"),
    origin: CoordinatesSchema,
    destination: CoordinatesSchema,
    optimize_for: ObjectiveSchema,
    return_alternatives: z
        .boolean()
        .default(true)
        .describe("Whether to show alternative routes"),
    raw_message: z.string().describe("The original user message verbatim"),
});

export const ParsedTSPRouteSchema = z.object({
    intent: z.literal("multi_destination"),
    origin: CoordinatesSchema,
    destinations: z
        .array(CoordinatesSchema)
        .min(2)
        .max(10)
        .describe("All destinations to visit"),
    optimize_for: ObjectiveSchema,
    return_to_origin: z
        .boolean()
        .default(false)
        .describe("Whether the route should loop back to the starting point"),
    raw_message: z.string().describe("The original user message verbatim"),
});

export const ParsedClarificationSchema = z.object({
    intent: z.literal("clarification_needed"),
    missing_fields: z
        .array(z.string())
        .describe("List of fields that could not be determined"),
    clarification_prompt: z
        .string()
        .describe("A friendly question to ask the user to fill the gaps"),
    raw_message: z.string().describe("The original user message verbatim"),
});

export const ParsedUserIntentSchema = z.discriminatedUnion("intent", [
    ParsedSingleRouteSchema,
    ParsedTSPRouteSchema,
    ParsedClarificationSchema,
]);

export type ParsedUserIntent = z.infer<typeof ParsedUserIntentSchema>;
export type ParsedSingleRoute = z.infer<typeof ParsedSingleRouteSchema>;
export type ParsedTSPRoute = z.infer<typeof ParsedTSPRouteSchema>;
export type ParsedClarification = z.infer<typeof ParsedClarificationSchema>;

// FastAPI Response Schemas

export const RouteLegSchema = z.object({
    from_stop_name: z.string(),
    to_stop_name: z.string(),
    agency_type: z.enum(["rail", "bus", "walk"]),
    edge_type: z.enum(["transit", "walk"]),
    departure_time: z.string().optional(),
    arrival_time: z.string().optional(),
    travel_time_sec: z.number(),
    fare_myr: z.number(),
    co2_grams: z.number(),
    distance_km: z.number(),
    route_id: z.string(),
    trip_id: z.string(),
});

export const CarbonSummarySchema = z.object({
    total_transit_co2_grams: z.number(),
    car_baseline_co2_grams: z.number(),
    co2_saved_grams: z.number(),
    co2_saved_percent: z.number(),
    equivalent_tree_days: z.number(),
    total_distance_km: z.number(),
    breakdown_by_mode: z.record(z.number()),
});

export const RouteResultSchema = z.object({
    legs: z.array(RouteLegSchema),
    total_time_sec: z.number(),
    total_fare_myr: z.number(),
    total_co2_grams: z.number(),
    total_distance_km: z.number(),
    car_co2_grams: z.number(),
    co2_saved_grams: z.number(),
    instructions: z.array(z.string()),
    objective: z.string(),
    carbon: CarbonSummarySchema.optional(),
});

export const RouteResponseSchema = z.object({
    recommended: RouteResultSchema.nullable(),
    alternatives: z.array(RouteResultSchema),
    reasoning: z.string(),
    origin_stop: z
        .object({ id: z.string(), name: z.string(), distance_m: z.number() })
        .optional(),
    destination_stop: z
        .object({ id: z.string(), name: z.string(), distance_m: z.number() })
        .optional(),
    error: z.string().optional().nullable(),
});

export const TSPResponseSchema = z.object({
    ordered_destinations: z.array(
        z.object({ lat: z.number(), lon: z.number() })
    ),
    legs_between_stops: z.array(RouteResultSchema),
    total_time_sec: z.number(),
    total_fare_myr: z.number(),
    total_co2_grams: z.number(),
    reasoning: z.string(),
    error: z.string().optional().nullable(),
});

export type RouteResponse = z.infer<typeof RouteResponseSchema>;
export type TSPResponse = z.infer<typeof TSPResponseSchema>;
export type RouteResult = z.infer<typeof RouteResultSchema>;

File: Backend/app/models/schemas.py
from typing import Any, Dict, List, Literal, Optional
from pydantic import BaseModel, Field


# Shared sub-models
class Coordinates(BaseModel):
    lat: float = Field(..., description="Latitude")
    lon: float = Field(..., description="Longitude")


class RouteLeg(BaseModel):
    from_stop_id: str
    from_stop_name: str
    to_stop_id: str
    to_stop_name: str
    trip_id: str
    route_id: str
    agency_type: str  # "rail" | "bus" | "walk"
    edge_type: str  # "transit" | "walk"
    departure_time: Optional[str] = None
    arrival_time: Optional[str] = None
    travel_time_sec: int
    fare_myr: float
    co2_grams: float
    distance_km: float


class CarbonSummary(BaseModel):
    total_transit_co2_grams: float
    car_baseline_co2_grams: float
    co2_saved_grams: float
    co2_saved_percent: float
    equivalent_tree_days: float
    total_distance_km: float
    breakdown_by_mode: Dict[str, float]
    emission_factors_used: Dict[str, float]


class RouteResult(BaseModel):
    legs: List[RouteLeg]
    total_time_sec: int
    total_fare_myr: float
    total_co2_grams: float
    total_distance_km: float
    car_co2_grams: float
    co2_saved_grams: float
    instructions: List[str]
    objective: str
    carbon: Optional[CarbonSummary] = None


class StopInfo(BaseModel):
    id: str
    name: str
    distance_m: float

# Route optimization request / response
class RouteRequest(BaseModel):
    origin: Coordinates
    destination: Coordinates
    optimize_for: Literal["time", "cost", "eco"] = Field(
        default="time", description="Optimization objective: 'time', 'cost', or 'eco'"
    )
    return_alternatives: bool = Field(
        default=True, description="Whether to return alternative routes"
    )


class RouteResponse(BaseModel):
    recommended: Optional[RouteResult]
    alternatives: List[RouteResult] = []
    reasoning: str
    origin_stop: Optional[StopInfo] = None
    destination_stop: Optional[StopInfo] = None
    error: Optional[str] = None


# TSP (multi-destination) request / response
class TSPRequest(BaseModel):
    origin: Coordinates
    destinations: List[Coordinates] = Field(
        ...,
        min_length=2,
        max_length=10,
        description="2–10 destinations to visit in the most optimal order",
    )
    optimize_for: Literal["time", "cost", "eco"] = Field(default="time")
    return_to_origin: bool = Field(
        default=False,
        description="Whether the route should return to the starting point",
    )


class TSPResponse(BaseModel):
    ordered_destinations: List[Coordinates]
    legs_between_stops: List[RouteResult]
    total_time_sec: int
    total_fare_myr: float
    total_co2_grams: float
    reasoning: str
    error: Optional[str] = None


#  Graph health / debug
class GraphStatsResponse(BaseModel):
    node_count: int
    edge_count: int
    feed_counts: Dict[str, int]

# Realtime
class VehiclePositionInfo(BaseModel):
    vehicle_id: str
    trip_id: str
    route_id: str
    lat: float
    lon: float
    timestamp: int
    current_stop_sequence: int
    current_status: str  # "INCOMING_AT" | "STOPPED_AT" | "IN_TRANSIT_TO"


class RealtimeStatusResponse(BaseModel):
    feeds: Dict[str, List[VehiclePositionInfo]]
    total_vehicles: int
